src/lib.rs

```diff
@@ -100,82 +112,284 @@ pub async fn run_cargo_command(
     );
 
     let mut cmd = Command::new(&cargo);
+    if is_run {
+        cmd.arg("build");
+        if let Some(arg) = args.peek() {
+            if Some((*arg).as_str()) == subcommand {
+                args.next().unwrap();
+            }
+        }
+    }
     cmd.args(args);
 
-    let is_build = matches!(subcommand, Some("b") | Some("build") | Some("rustc"));
+    // We need to set the `color` parameter to "always", otherwise
+    // ANSI colors are being stripped out by default
+    if !build_args
+        .iter()
+        .any(|a| a == "--color" || a.starts_with("--color="))
+    {
+        cmd.arg("--color").arg("always");
```

Comment:
As it appears that only `stderr` has colorized output when `--message-format json-render-diagnostics` is used, could we not use `.stdout(piped())` and get a reader which processes each line of JSON output, buffering only what wasm artifacts were built as it goes? Thus `stderr` would not be redirected, so we don't need to muck with `--color` and let it auto-detect a tty.

This would have the added benefit (and a complete deal breaker if not implemented) of not buffering the `stderr` output, making it look like `cargo-component` is hanging until the build completes and it blasts the entire output.

`cargo-component` is meant to look and feel like cargo itself, and if commands like `build` and `run` can't give immediate build progress output, then we need to find an alternative implementation.

Reply:
I believe my first try was with a piped stdout and stderr but I ran into some complexity that I don't recall right now. I will make the modification and check if the approach you are suggesting will be more performant. I assume that performance/speed is what you were expecting to hurt DX, right?

Reply:
Sorry I missed this question from two weeks ago (I was out last week too).

I'm not _as_ concerned about the performance (although that is very important) as I am about buffering the output of the spawned processes, which makes it appear like nothing is happening while cargo is running.